packages/block-library/src/navigation/view.js

```diff
@@ -18,205 +18,172 @@ const focusableSelectors = [
+			)
+				actions.openMenu( 'hover' );
+		},
+		closeMenuOnHover() {
+			actions.closeMenu( 'hover' );
+		},
+		openMenuOnClick() {
+			const ctx = getContext();
+			const { ref } = getElement();
+			ctx.previousFocus = ref;
+			actions.openMenu( 'click' );
+		},
+		closeMenuOnClick() {
+			actions.closeMenu( 'click' );
+			actions.closeMenu( 'focus' );
+		},
+		openMenuOnFocus() {
+			actions.openMenu( 'focus' );
+		},
+		toggleMenuOnClick() {
+			const ctx = getContext();
+			const { ref } = getElement();
+			// Safari won't send focus to the clicked element, so we need to manually place it: https://bugs.webkit.org/show_bug.cgi?id=22261
+			if ( window.document.activeElement !== ref ) ref.focus();
+			const { menuOpenedBy } = state;
+			if ( menuOpenedBy.click || menuOpenedBy.focus ) {
+				actions.closeMenu( 'click' );
+				actions.closeMenu( 'focus' );
+			} else {
+				ctx.previousFocus = ref;
+				actions.openMenu( 'click' );
+			}
+		},
+		handleMenuKeydown( event ) {
+			const { type, firstFocusableElement, lastFocusableElement } =
+				getContext();
+			if ( state.menuOpenedBy.click ) {
+				// If Escape close the menu.
+				if ( event?.key === 'Escape' ) {
+					actions.closeMenu( 'click' );
+					actions.closeMenu( 'focus' );
+					return;
+				}
 
-					// The event.relatedTarget is null when something outside the navigation menu is clicked. This is only necessary for Safari.
+				// Trap focus if it is an overlay (main menu).
+				if ( type === 'overlay' && event.key === 'Tab' ) {
+					// If shift + tab it change the direction.
 					if (
-						event.relatedTarget === null ||
-						( ! context.core.navigation.modal?.contains(
-							event.relatedTarget
-						) &&
-							event.target !== window.document.activeElement )
+						event.shiftKey &&
+						window.document.activeElement === firstFocusableElement
 					) {
-						closeMenu( store, 'click' );
-						closeMenu( store, 'focus' );
+						event.preventDefault();
+						lastFocusableElement.focus();
+					} else if (
+						! event.shiftKey &&
+						window.document.activeElement === lastFocusableElement
+					) {
+						event.preventDefault();
+						firstFocusableElement.focus();
 					}
-				},
-			},
+				}
+			}
+		},
+		handleMenuFocusout( event ) {
+			const { modal } = getContext();
+			// If focus is outside modal, and in the document, close menu
+			// event.target === The element losing focus
+			// event.relatedTarget === The element receiving focus (if any)
+			// When focusout is outsite the document,
+			// `window.document.activeElement` doesn't change.
+
+			// The event.relatedTarget is null when something outside the navigation menu is clicked. This is only necessary for Safari.
+			if (
+				event.relatedTarget === null ||
+				( ! modal?.contains( event.relatedTarget ) &&
+					event.target !== window.document.activeElement )
+			) {
+				actions.closeMenu( 'click' );
+				actions.closeMenu( 'focus' );
+			}
+		},
+
+		openMenu( menuOpenedOn ) {
```

Comment:
Maybe we can add a default so it can be referenced in directives.

```suggestion
		openMenu( menuOpenedOn = 'click' ) {
```

Reply:
Done in https://github.com/WordPress/gutenberg/pull/55459/commits/9869b642dd0a23c9016a08f4398fa778faac7cd0.